### ISSUE TITLE
obtain '.mat' semantic segmentation files from this code instead of 'jpg' files. Here I get the output '.mat' file as a 4-D Uint8 output. Is there a way to get a 2-D int32 output ".mat" file.

### ISSUE THREAD

**Issue by Ravindu-Yasas-Nagasinghe** [chaofengc/PSFRGAN#53]
Here I get the output '.mat' file as a 4-D Uint8 output. Is there a way to get a 2-D int32 output ".mat" file.
given below is the code for my output. the part for getting the '.mat' file is in the last 8 -13 lines

`
'''
This script enhance all faces in one image with PSFR-GAN and paste it back to the original place.
'''
import dlib
import os
import cv2
import numpy as np
from tqdm import tqdm
from skimage import transform as trans
from skimage import io

import torch
from utils import utils
from options.test_options import TestOptions
from models import create_model
from scipy.io import savemat

def detect_and_align_faces(img, face_detector, lmk_predictor, template_path, template_scale=2, size_threshold=999):
align_out_size = (512, 512)
ref_points = np.load(template_path) / template_scale

# Detect landmark points
face_dets = face_detector(img, 1)
assert len(face_dets) > 0, 'No faces detected'

aligned_faces = []
tform_params = []
for det in face_dets:
    if isinstance(face_detector, dlib.cnn_face_detection_model_v1):
        rec = det.rect # for cnn detector
    else:
        rec = det
    if rec.width() > size_threshold or rec.height() > size_threshold: 
        print('Face is too large')
        break
    landmark_points = lmk_predictor(img, rec) 
    single_points = []
    for i in range(5):
        single_points.append([landmark_points.part(i).x, landmark_points.part(i).y])
    single_points = np.array(single_points)
    tform = trans.SimilarityTransform()
    tform.estimate(single_points, ref_points)
    tmp_face = trans.warp(img, tform.inverse, output_shape=align_out_size, order=3)
    aligned_faces.append(tmp_face*255)
    tform_params.append(tform)
return [aligned_faces, tform_params]
def def_models(opt):
model = create_model(opt)
model.load_pretrain_models()
model.netP.to(opt.device)
model.netG.to(opt.device)
return model

def enhance_faces(LQ_faces, model):
hq_faces = []
lq_parse_maps = []
for lq_face in tqdm(LQ_faces):
with torch.no_grad():
lq_tensor = torch.tensor(lq_face.transpose(2, 0, 1)) / 255. * 2 - 1
lq_tensor = lq_tensor.unsqueeze(0).float().to(model.device)
parse_map, _ = model.netP(lq_tensor)
parse_map_onehot = (parse_map == parse_map.max(dim=1, keepdim=True)[0]).float()
output_SR = model.netG(lq_tensor, parse_map_onehot)
hq_faces.append(utils.tensor_to_img(output_SR))
lq_parse_maps.append(utils.color_parse_map(parse_map_onehot)[0])
return hq_faces, lq_parse_maps

def past_faces_back(img, hq_faces, tform_params, upscale=1):
h, w = img.shape[:2]
img = cv2.resize(img, (int(wupscale), int(hupscale)), interpolation=cv2.INTER_CUBIC)
for hq_img, tform in tqdm(zip(hq_faces, tform_params), total=len(hq_faces)):
tform.params[0:2,0:2] /= upscale
back_img = trans.warp(hq_img/255., tform, output_shape=[int(hupscale), int(wupscale)], order=3) * 255

    # blur mask to avoid border artifacts
    mask = (back_img == 0) 
    mask = cv2.blur(mask.astype(np.float32), (5,5))
    mask = (mask > 0)
    img = img * mask + (1 - mask) * back_img 
return img.astype(np.uint8)
def save_imgs(img_list, save_dir):
for idx, img in enumerate(img_list):
save_path = os.path.join(save_dir, '{:03d}.jpg'.format(idx))
io.imsave(save_path, img.astype(np.uint8))

if name == 'main':
opt = TestOptions().parse()
# face_detector = dlib.get_frontal_face_detector()
face_detector = dlib.cnn_face_detection_model_v1('./pretrain_models/mmod_human_face_detector.dat')
lmk_predictor = dlib.shape_predictor('./pretrain_models/shape_predictor_5_face_landmarks.dat')
template_path = './pretrain_models/FFHQ_template.npy'

print('======> Loading images, crop and align faces.')
img_path = opt.test_img_path 
img = dlib.load_rgb_image('/home/zebra/Deep3dPortrait/examples/1_1234.jpg')
aligned_faces, tform_params = detect_and_align_faces(img, face_detector, lmk_predictor, template_path)
# Save aligned LQ faces
save_lq_dir = os.path.join(opt.results_dir, 'LQ_faces') 
os.makedirs(save_lq_dir, exist_ok=True)
print('======> Saving aligned LQ faces to', save_lq_dir)
save_imgs(aligned_faces, save_lq_dir)

enhance_model = def_models(opt)
hq_faces, lq_parse_maps = enhance_faces(aligned_faces, enhance_model)
# Save LQ parsing maps and enhanced faces
save_parse_dir = os.path.join(opt.results_dir, 'ParseMaps') 
save_hq_dir = os.path.join(opt.results_dir, 'HQ') 
os.makedirs(save_parse_dir, exist_ok=True)
os.makedirs(save_hq_dir, exist_ok=True)
print('======> Save parsing map and the enhanced faces.')
arr = np.array(lq_parse_maps)
print(arr)
mdic = {"lq_parse_maps": lq_parse_maps}
savemat("custom_input.mat", mdic)
print('######################################')
print(mdic)
save_imgs(lq_parse_maps, save_parse_dir)
save_imgs(hq_faces, save_hq_dir)

print('======> Paste the enhanced faces back to the original image.')
hq_img = past_faces_back(img, hq_faces, tform_params, upscale=opt.test_upscale) 
final_save_path = os.path.join(opt.results_dir, 'hq_final.jpg') 
print('======> Save final result to', final_save_path)
io.imsave(final_save_path, hq_img)

**Comment by syysummer** [comment]
> 在这里，我将输出“.mat”文件作为 4-D Uint8 输出。有没有办法获得二维 int32 输出“.mat”文件。 下面给出的是我的输出的代码。获取“.mat”文件的部分位于最后 8 -13 行
> 
> ` ''' 该脚本使用 PSFR-GAN 增强一张图像中的所有面孔，并将其粘贴回原始位置。 ''' 导入 dlib 导入 o​​s 导入 cv2从 tqdm 导入 numpy 作为 np 从 skimage 导入 tqdm导入转换为 trans 从 skimage 导入 io
> 
> 从 utils导入 torch 从 options.test_options 导入 utils 从模型导入 TestOptions 从 scipy.io 导入 create_model 导入 savemat
> 
> def detector_and_align_faces（img，face_ detector，lmk_predictor，template_path，template_scale = 2，size_threshold = 999）： align_out_size =（512，512） ref_points = np.load（template_path）/ template_scale
> 
> # 检测标志点
> face_dets = Face_Detector(img, 1) assert len(face_dets) > 0, '未检测到人脸'
> 
> aligned_faces=[] tform_params=[] fordetinface_dets: ifisinstance(face_Detector,dlib.cnn_face_detection_model_v1): rec=det.rect#对于cnn检测器 else: rec= detifrec.width()>size_threshold或rec.height( ) > size_threshold: print('人脸太大') breaklandmark_points = lmk_predictor(img, rec) single_points = [] for i in range(5): single_points.append([landmark_points.part(i).x,landmark_points.部分(i).y]) single_points = np.array(single_points) tform = trans.SimilarityTransform() tform.estimate(single_points, ref_points) tmp_face = trans.warp(img, tform.inverse, output_shape=align_out_size, order=3 ） aligned_faces.append（tmp_face * 255） tform_params.append（tform） 返回[aligned_faces，tform_params] def def_models（opt）： 模型= create_model（opt） model.load_pretrain_models（） model.netP.to（opt.device） 模型。 netG.to(opt.device) 返回模型
> 
> def Enhance_faces(LQ_faces, model): hq_faces = [] lq_parse_maps = [] for lq_face in tqdm(LQ_faces): with torch.no_grad(): lq_tensor = torch.tensor(lq_face.transpose(2, 0, 1)) / 255 . * 2 - 1 lq_tensor = lq_tensor.unsqueeze(0).float().to(model.device) parse_map, _ = model.netP(lq_tensor) parse_map_onehot = (parse_map == parse_map.max(dim=1, keepdim= True)[0]).float() output_SR = model.netG(lq_tensor, parse_map_onehot) hq_faces.append(utils.tensor_to_img(output_SR)) lq_parse_maps.append(utils.color_parse_map(parse_map_onehot)[0]) 返回 hq_faces, lq_parse_maps
> 
> def past_faces_back(img, hq_faces, tform_params, upscale=1): h, w = img.shape[:2] img = cv2.resize(img, (int(wupscale), int(hupscale)), interpolation=cv2.INTER_CUBIC ） 对于 hq_img，tqdm 中的 tform（zip（hq_faces，tform_params），total=len（hq_faces））： tform.params[0:2,0:2] /= 高档 back_img = trans.warp（hq_img/255., tform ，output_shape=[int(hupscale)，int(wupscale)]，order=3) * 255
> 
> ```
> # blur mask to avoid border artifacts
> mask = (back_img == 0) 
> mask = cv2.blur(mask.astype(np.float32), (5,5))
> mask = (mask > 0)
> img = img * mask + (1 - mask) * back_img 
> ```
> 
> return img.astype(np.uint8) def save_imgs(img_list, save_dir): 对于 idx, img in enumerate(img_list): save_path = os.path.join(save_dir, '{:03d}.jpg'.format(idx) ) io.imsave(save_path, img.astype(np.uint8))
> 
> if name == 'main': opt = TestOptions().parse()
> 
> # Face_Detector = dlib.get_frontal_face_Detector()
> Face_Detector = dlib.cnn_face_detection_model_v1('./pretrain_models/mmod_ human_face_detector.dat') lmk_predictor = dlib.shape_predictor('./pretrain_models/shape_predictor_5_face_landmarks.dat') template_path = './pretrain_models/FFHQ_template.npy'
> 
> print('======> 加载图像，裁剪并对齐脸部。') img_path = opt.test_img_path img = dlib.load_rgb_image('/home/zebra/Deep3dPortrait/examples/1_1234.jpg') aligned_faces, tform_params = detector_and_align_faces（img，face_ detector，lmk_predictor，template_path）
> 
> # 保存对齐的 LQ 面
> save_lq_dir = os.path.join(opt.results_dir, 'LQ_faces') os.makedirs(save_lq_dir, exit_ok=True) print('======> 将对齐的 LQ 面保存到', save_lq_dir) save_imgs(aligned_faces, save_lq_dir) ）
> 
> 增强模型 = def_models(opt) hq_faces, lq_parse_maps = 增强_faces(aligned_faces, 增强_模型)
> 
> # 保存 LQ 解析图和增强的人脸
> save_parse_dir = os.path.join(opt.results_dir，'ParseMaps') save_hq_dir = os.path.join(opt.results_dir，'HQ') os.makedirs(save_parse_dir，exist_ok=True) os.makedirs(save_hq_dir，exist_ok= True) print('======> 保存解析图和增强的面。') arr = np.array(lq_parse_maps) print(arr) mdic = {"lq_parse_maps": lq_parse_maps} savemat("custom_input.mat" , mdic) 打印('########################################') 打印(mdic) save_imgs（lq_parse_maps，save_parse_dir） save_imgs（hq_faces，save_hq_dir）
> 
> print('======> 将增强后的人脸粘贴回原始图像。') hq_img = past_faces_back(img, hq_faces, tform_params, upscale=opt.test_upscale) Final_save_path = os.path.join(opt.results_dir, 'hq_final.jpg') print('======> 将最终结果保存到', Final_save_path) io.imsave(final_save_path, hq_img)

Thank you so much! I can generate .mat files using the method you provided, but when applied to Deep3dPortrait, an error message "operands could not be broadcast together with shapes (256,256,3) (256,256)" will appear. Have you solved this problem? How can I generate the correct .mat file?
